### ISSUE TITLE
Recursive dependency on Carp

### ISSUE THREAD

**Issue by cpansprout** [Test-More/test-more#682]
I know I’m a troglodyte, but I’m trying to run some simple module tests on perl 5.8.7 (!!!) to track down a CPAN Testers failure.

I keep running into a problem with CPAN.pm complaining about a recursive dependency and then refusing to do anything (this is CPAN 1.76_01).

Rather than forcing users to upgrade their CPAN client, would you be willing to drop the dependency on Carp?  Carp has been included with perl since perl 5.000, and is highly unlikely ever to be removed, so specifying it as a dependency is not strictly necessary.

Test::More is low enough in the dependency chain (it’s practically a bootstrapping module), so I think it makes sense to eschew pedantic correctness in this instance.

**Comment by exodist** [comment]
I set the dependency to Carp version 0, so I am surprised it is trying to upgrade. Does carp have no version number in older perls?

In any case it seems silly to block you on this. I would be fine with removing the listing, I will try to get a release out for you some time this week. I am away from my computer right now (using phone)

**Comment by karenetheridge** [comment]
Carp has had a version since perl 5.7.3.

@cpansprout what is the recursive dependency? Presumably it involves Carp itself?

**Comment by cpansprout** [comment]
> I set the dependency to Carp version 0, so I am surprised it is trying to upgrade. Does carp have no version number in older perls?

To answer your question, 5.7.3 was the first perl version in which Carp had a version number (1.01).

But it’s not that.  It’s just that the older CPAN client is paranoid about recursive dependencies, so I can’t upgrade both of them in the same session.  As soon as it detects the circularity, it freaks out and won‘t even let me `look Carp`.

(To be precise, though, I am not actually installing them, but running `test Test::More Carp` in order to test modules with the newer versions.  Having older versions installed is useful for finding bugs like <https://rt.cpan.org/Ticket/Display.html?id=115314> and <https://rt.cpan.org/Ticket/Display.html?id=115312>.)

> @cpansprout what is the recursive dependency? Presumably it involves Carp itself?
I think it would be better to resolve the underlying issue rather than omit prerequisites from being declared.

Test::More depends on Carp and Carp depends on Test::More.  Likewise, I have just found that Test::More depends on Scalar::Util and Scalar::Util depends on Test::More.

Again, Scalar::Util has been in core since 5.7.3.  The version that Test::More requires is 1.13, which has been in core since 5.8.1, the earliest version that Test::More runs on now.

Because Test::More is so ancient and venerable, any new dependency that it gains is likely to depend on Test::More itself, so I don’t know the real solution to this.

(Maybe the *real* solution is for me to drop support for 5.8 in my modules. :-)

**Comment by karenetheridge** [comment]
> (Maybe the real solution is for me to drop support for 5.8 in my modules. :-)

Is it possible for you to upgrade your cpan client instead? That would be a little less drastic.

**Comment by cpansprout** [comment]
> Is it possible for you to upgrade your cpan client instead? That would be a little less drastic.

I suppose so.  I will have to force install (or just do it by hand, which might be easier).

I guess I’ll just go ahead and close this.

**Comment by cpansprout** [comment]
I was mistaken.  With the *latest* CPAN (2.10), this is still a problem.
```
cpan> test Test::More File::Spec
...
cpan> install Test::More

Recursive dependency detected:
    Test::More (have: 0.54; want: 1.302026)
 => EXODIST/Test-Simple-1.302026.tar.gz
 => File::Spec (have: 3.05; requires: 0)
 => RJBS/PathTools-3.62.tar.gz
 => Test::More (have: 0.54; requires: 0.88).
Cannot resolve.
...
```

Since Test::More is so low-level (*everything* depends on it, even the modules it uses), I would suggest removing from its prerequisites all modules that were in core in 5.8.1, which happens to be all of its prerequisites.

**Comment by karenetheridge** [comment]
This looks like a CPAN bug -- PathTools is already at 3.05, so that prereq is satisfied. There is no need to fetch a newer PathTools.

**Comment by cpansprout** [comment]
Is it really worth making CPAN.pm more complicated?  In any case, I have opened <https://rt.cpan.org/Ticket/Display.html?id=115340> against CPAN.pm.  (That does not mean the issue is resolved.  I still think not declaring core-only dependency would be the easier and simpler solution.)

**Comment by karenetheridge** [comment]
We've been trying to avoid under-declaring prerequisites, as various static inspection tools want to know this information (for example for building dependency trees of the toolchain).

**Comment by exodist** [comment]
I do not think there is a clear "correct" answer here. At this point I think that there are valid reasons for each way, and I need to pick one. My choice is to leave the core deps listed.

**Comment by haarg** [comment]
So you are favoring breaking a legitimate use case to enable a basically nonexistent use case?

**Comment by exodist** [comment]
I am choosing not to enter into the hell of trying to please everyone all the time. Everyone has an opinion on this, no opinion is obviously superior unless you are the one holding the opinion. At this point I need to pick a direction to end the argument. I picked the direction I did because it is my preference (by no means a strong preference).

Feel free to keep debating, but at this point I don't consider this an issue worth arguing over and I don't intend to reply further unless someone can give a convinsing technical argument with clear benefit to a clear majority of users. At this point I see a very small minority for each camp and it is likely that the vast majority of users don't care. Not worth the energy.

**Comment by exodist** [comment]
Since @rjbs and @xdg are right next to me I asked their feedback. The debate was almost a mirror of this thread. However there was a clarificarion. This ticket only appears to reference Carp, and ask that the Carp module be unlisted, it is not asking for ALL core deps to be unlisted. This is because Carp is a special case since it had no version number in older perls.

With this I will remove the Carp listing, but will not remove the other core dep listings.

**Comment by karenetheridge** [comment]
Ultimately, the CPAN spec says that all prerequisites can and should be listed.  All tools should be compliant with the spec.

However, the fact that some tools are broken in various ways may make us take actions to mitigate the real-world effects of this, but we should try to minimize these.

tldr: CPAN.pm is still broken, no matter what we do in this distribution.
